File: 2015/08/rust/src/main.rs
use std::fs::File;
use std::io::{ BufRead, BufReader };

// --- Constants --
const INPUT_FILE: &'static str = "../input/input.txt";

fn count_mem(s: &str) -> usize {
    let mut total = s.len() - 2;
    let a: Vec<_> = s.chars().collect();
    let mut i = 1;
    while i < a.len() - 2 {
        if a[i] == '\\' {
            if a[i + 1] == '\\' || a[i + 1] == '"' {
                total -= 1;
                i += 1;
            } else if a[i + 1] == 'x' {
                total -= 3;
                i += 3;
            }
        }
        i += 1;
    }
    total
}

fn encode(s: &str) -> String {
    let mut v: Vec<char> = Vec::new();
    v.push('"');
    for c in s.chars() {
        if c == '"' || c == '\\' {
            v.push('\\');
        }
        v.push(c);
    }
    v.push('"');
    v.into_iter().collect()
}

// --- Main function ---
fn main() -> Result<(), Box<dyn std::error::Error>> {
    let mut total1 = 0;
    let mut total2 = 0;

    // --- Read and parse the input file ---
    let file = File::open(INPUT_FILE)?;
    for line in BufReader::new(file).lines() {
        let line = line.unwrap();
        total1 += line.len() - count_mem(&line);

        let enc = encode(&line);
        total2 += enc.len() - count_mem(&enc);
    }

    // --- Puzzle 1 ---
    println!("1. Number of characters in memory: {}", total1);

    // --- Puzzle 2 ---
    println!("2. Number of characters in memory: {}", total2);

    Ok(())
}

File: 2015/08/go/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
)

// --- Constants ---
const INPUT_FILE = "../input/input.txt"

// Calcuates the used memory from a string
func countMem(s string) int {
	total := len(s) - 2
	i := 1
	for i < len(s)-2 {
		if s[i] == '\\' {
			if i+1 < len(s) && (s[i+1] == '\\' || s[i+1] == '"') {
				total--
				i++
			} else if i+2 < len(s) && s[i+1] == 'x' {
				total -= 3
				i += 3
			}
		}
		i++
	}
	return total
}

// Encode the string
func encode(s string) string {
	var ls []string
	ls = append(ls, `"`)
	for _, c := range s {
		if c == '"' || c == '\\' {
			ls = append(ls, "\\")
		}
		ls = append(ls, string(c))
	}
	ls = append(ls, `"`)
	return strings.Join(ls, "")
}

func main() {
	// --- Variables ---
	total1 := 00
	total2 := 0

	// --- Read input file ---
	f, err := os.Open(INPUT_FILE)
	if err != nil {
		fmt.Println("Error al abrir archivo:", err)
		return
	}
	defer f.Close()

	// --- Parse each line ---
	scanner := bufio.NewScanner(f)
	for scanner.Scan() {
		line := scanner.Text()

		// --- Puzzle 1 ---
		mem := countMem(line)
		total1 += len(line) - mem

		// --- Puzzle 2 ---
		enc := encode(line)
		total2 += len(enc) - countMem(enc)
	}

	// --- Print results ---
	fmt.Printf("1. Number of characters in memory: %d\n", total1)
	fmt.Printf("2. Number of characters in memory: %d\n", total2)
}
